test/core.jl

```diff
@@ -3499,3 +3499,10 @@ let
     finalize(obj)
     @test finalized == 1
 end
+
+# non-terminating inference, issue #14009
+type A14009{T}
+end
+A14009{T}(a::T) = A14009{T}()
+f14009(a) = ? ? f14009(A14009(a)) : a
```

Comment:
Any less confusing way of specifying a condition that type inference cannot reason about?